generator/schemas/gen.go

```diff
@@ -0,0 +1,308 @@
+			return nil
+		}
+
+		if len(jsonschemaRequested) == 0 {
+			return nil
+		}
+
+		var devfileSchemaVersion *semver.Version
+		packageMarkers, err := markers.PackageMarkers(ctx.Collector, root)
+		if err != nil {
+			return err
+		}
+		switch markerValue := packageMarkers.Get(jsonschemaVersionMarker.Name).(type) {
+		case string:
+			devfileSchemaVersion, err = semver.NewVersion(markerValue)
+			if err != nil {
+				root.AddError(fmt.Errorf("In order to generate Json schemas from the K8S API, you should provide a valid semver-compatible devfile version in the +devfile:jsonschema:version comment marker of the K8S API package (in the doc.go file)"))
+				return nil
+			}
+		default:
+			root.AddError(fmt.Errorf("In order to generate Json schemas from the K8S API, you should annotate the K8S API package (in the doc.go file) with the +devfile:jsonschema:version comment marker"))
+			return nil
+		}
+
+		for _, typeToProcess := range jsonschemaRequested {
+			typeIdent := crd.TypeIdent{
+				Package: root,
+				Name:    typeToProcess.Name,
+			}
+			schemaGenerateMarker := typeToProcess.Markers.Get(jsonschemaGenerateMarker.Name).(GenerateJSONSchema)
+			parser.NeedFlattenedSchemaFor(typeIdent)
+			currentJSONSchema, found := parser.FlattenedSchemata[typeIdent]
+			if !found {
+				root.AddError(fmt.Errorf("Json schema for type " + typeIdent.Package.Name + "/" + typeIdent.Name + " could not be generated"))
+				continue
+			}
+
+			fieldsToSkip := []string{}
+			if schemaGenerateMarker.OmitCustomUnionMembers {
+				fieldsToSkip = append(fieldsToSkip, "Custom")
+			}
+
+			genutils.AddUnionOneOfConstraints(&currentJSONSchema, unionDiscriminators, true, fieldsToSkip...)
+
+			// Fix descriptions to have them Markdown compatible
+			genutils.EditJSONSchema(&currentJSONSchema, func(schema *apiext.JSONSchemaProps) (newVisitor genutils.Visitor, stop bool) {
+				if schema == nil || schema.Description == "" {
+					return
+				}
+				regex, err := regexp.Compile(" \\n ([^-])")
+				if err != nil {
+					root.AddError(err)
+					return
+				}
+				schema.Description = strings.ReplaceAll(schema.Description, " \t", "\n")
+				schema.Description = strings.ReplaceAll(schema.Description, " \n - ", "\n- ")
+				schema.Description = regex.ReplaceAllString(schema.Description, "\n\n$1")
+				return
+			})
+
+			// Add the additionalProperies required to reflect the expected behavior from the K8S API,
+			// (preserve-unknown-fields false by default unless on the Devfile metadata)
+			genutils.EditJSONSchema(&currentJSONSchema, func(schema *apiext.JSONSchemaProps) (newVisitor genutils.Visitor, stop bool) {
+				if schema == nil ||
+					schema.Type != "object" ||
+					schema.AdditionalProperties != nil {
+					return
+				}
+				schema.AdditionalProperties = &apiext.JSONSchemaPropsOrBool{
+					Allows: schema.XPreserveUnknownFields != nil && *schema.XPreserveUnknownFields,
+				}
+				return
+			})
+
+			// Fix descriptions to have them Markdown compatible
```

Comment:
There's an identical comment above (line 159); why can't we do both at the same time?

Reply:
The comment is just wrong copy-paste. Let me fix it.

Reply:
Fixed in commit https://github.com/devfile/api/pull/133/commits/5deb2bcccdeac386f390562308b7f6095847b199

Reply:
Fixed in commit https://github.com/devfile/api/pull/133/commits/dc2f2e328fdcc52f55898f028ef900277015f387